### ISSUE TITLE
Module request: Port the `nix` modules from NixOS

### ISSUE THREAD

**Issue by ShamrockLee** [nix-community/home-manager#2324]
### Description

The Nix package manager offers unprivileged users the ability to customize its behavior through configurations. Since the configuration can be done declaratively on NixOS, porting them to Home Manager would be beneficial to users.

Example use cases:

*   Use nixUnstable with flakes locally
`home.nix`
```Nix
{config, lib, pkgs, ...}:
{
  nix.enable = true;
  nix.package = pkgs.nixUnstable;
  nix.extraOptions = ''
    experimental-features = nix-command flakes
  '';
}
```

*   Custom registries

`flake.nix`
```Nix
{
  inputs.nixpkgs.url = "github:NixOS/nixpkgs/nixos-20.09";
  inputs.nixpkgs-nixos-unstable.url = "github:NixOS/nixpkgs/nixos-unstable";
  inputs.home-manager = "github:nix-community/home-manager/release-20.09";
  inputs.home-manager.inputs.nixpkgs.follows = "nixpkgs";

  outputs = inputs@{ self, home-manager, ... }:
    let
      system = "x86_64-linux";
      username = "shamrock";
    in {
      homeManagerConfiguration.${username} = home-manager.lib.homeManagerConfiguration {
        configuration = import ./home.nix;
        inherit system username;
        homeDirectory = "/home/${username}";
        extraSpecialArgs = {
          # Inject the flakes here
          inherit (inputs) nixpkgs nixpkgs-nixos-unstable system-config-flake;
        };
      };
    };
}
```
`home.nix`
```Nix
args@{config, lib, pkgs, ... }: {
  nix = {
    enable = true;
    # Use the nix command inside the environment
    package = null;
    registry.local = {
      from = { type = "indirect"; id = "nixpkgs-local"; };
      flake = args.nixpkgs;
    };
    registry.unstable = {
      from = { type = "indirect"; id = "nixpkgs-local-unstable"; };
      flake = args.nixpkgs-unstable;
    };
  };
  # ...
}
```

    and then we can `nix search` the local nixpkgs input (without network connection) and `nix run` a program.
```console
$ nix search nixpkgs-local hello
* legacyPackages.x86_64-linux.gnome.iagno (3.38.1)
  Computer version of the game Reversi, more popularly called Othello

* legacyPackages.x86_64-linux.hello (2.10)
  A program that produces a familiar, friendly greeting

* legacyPackages.x86_64-linux.hello-unfree (1.0)
  An example package with unfree license (for testing)

* legacyPackages.x86_64-linux.hello-wayland (2020-07-27)
  Hello world Wayland client

* legacyPackages.x86_64-linux.javaPackages.mavenHello_1_0 (1.0)
  Maven Hello World

* legacyPackages.x86_64-linux.javaPackages.mavenHello_1_1 (1.1)
  Maven Hello World

$ nix run nixpkgs-local hello
Hello, world!

```

https://discourse.nixos.org/t/local-flake-based-nix-search-nix-run-and-nix-shell/13433/4

**Comment by berbiche** [comment]
According to the wiki [0], the user registry is configurable and written to `~/.config/nix/registry.json`.
The user nix.conf is stored in `$XDG_CONFIG_HOME/nix/nix.conf` or `$HOME/.config/nix/nix.conf` [1][2].

0: https://nixos.org/manual/nix/unstable/command-ref/new-cli/nix3-registry.html
1: https://nixos.org/manual/nix/unstable/command-ref/env-common.html
2: https://github.com/NixOS/nix/issues/2704

---

I imagine this could be very practical for people using Nix outside of nix-darwin and NixOS :+1:

**Comment by ShamrockLee** [comment]
As a fundamental part of NixOS, the nix options contains a large amount of sub-options. We would need to decide which to port, for there might be some operating-system-level stuff that doesn't make much sense for Home Manager users.

**Comment by polykernel** [comment]
I will take a shot at this if I have time, does nix.conf have any options which are not a key-value mapping? If not, then this can be trivially implemented as a [structural settings](https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md).

**Comment by ShamrockLee** [comment]
> does nix.conf have any options which are not a key-value mapping? If not, then this can be trivially implemented as a [structural settings](https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md).

Sounds great!

According to [`nixos/nix-daemon`](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/misc/nix-daemon.nix), all the configurations are in the form `key = value`. How about opening a PR in [NixOS/nixpkgs](https://github.com/NixOS/nixpkgs) for the structural-setting implementation and port it here after merged? That would keep both sides consistent.

**Comment by terlar** [comment]
Would still be nice to have this, I think their were some PR adding some brief nix config

**Comment by berbiche** [comment]
I suppose once https://github.com/NixOS/nixpkgs/pull/139075 is merged, work by @polykernel could be adapted to be used in Home Manager.

I wonder what options nix ignores from the user nix.conf file.

**Comment by ShamrockLee** [comment]
Considering the substantial changes made in https://github.com/NixOS/nixpkgs/pull/139075 , it would be good to wait a little while for it to stabilize.

**Comment by ShamrockLee** [comment]
Closed by mistake. Reopened.

**Comment by polykernel** [comment]
One issues that needs to be addressed in the context of the HM module is that Nix metadata(version) somehow needs to be passed into the module for config generation.

In NixOS, this can be retrieved from the package option given it is defined at the system level. In Home Manager, my idea has been define an option called `nixVersion` which specified the Nix version the config should be generated against but it forces the user to manually synchronize this version option with the system nix version.

It seems to me that manual intervention can not be avoided, does anyone have ideas on alternatives?

**Comment by ShamrockLee** [comment]
That sounds fair enough.

The HM module should give users the freedom to specify `nix.package = null;`, while users who do so must specify `nix.nixVersion`.

If one want to automate the process while using the Nix package from the system, she/he can specify the system configuration flake as a flake input and feed it and `hostname` to HM through `extraSpecialArgs`, and then do something like
```Nix
{ config, lib, pkgs, hostname, system-config-flake, ... }:
let
  config-system = system-config-flake.outputs.nixosConfigurations.${hostname}.config;
in
{
  nix.nixVersion = lib.getVersion config-system.nix.package;
}
```

**Comment by polykernel** [comment]
> The HM module should give users the freedom to specify `nix.package = null;`, while users who do so must specify `nix.nixVersion`.

I think `nix.package` should suffice because the config validation depends on the host platform Nix binary for `nix show-config`. The version is only used to determine appropriate flags to pass to the binary, which should be retrieved from the binary itself.

> If one want to automate the process while using the Nix package from the system, she/he can specify the system configuration flake as a flake input and feed it and `hostname` to HM through `extraSpecialArgs`

You are right, I believe it is also possible without flakes using `nixos-config` in the search path but I have not tested it.

**Comment by ShamrockLee** [comment]
This issue is now resolvod.
